src/util/CurrencyInfoHelpers.ts

```diff
@@ -14,6 +14,57 @@ export function isKeysOnlyPlugin(pluginId: string): boolean {
   return keysOnlyMode || ENV.KEYS_ONLY_PLUGINS[pluginId]
 }
 
+export type FindTokenParams =
+  | {
+      account: EdgeAccount
+      pluginId: string
+      networkLocation: JsonObject
+    }
+  | {
+      allTokens: EdgeTokenMap
+      networkLocation: JsonObject
+    }
+
+export const findTokenIdByNetworkLocation = (params: FindTokenParams): EdgeTokenId | undefined => {
+  const { networkLocation } = params
+  let allTokens: EdgeTokenMap
+  if ('allTokens' in params) {
+    allTokens = params.allTokens
+  } else {
+    allTokens = params.account.currencyConfig[params.pluginId]?.allTokens
+    if (allTokens == null) return
+  }
+
+  // Get the first tokenId. This could use Object.keys(allTokens)[0] but since we might have 1000s of
+  // tokens and this routines may get used in a LOT of places, I want to avoid interating over all tokens
+  // unnecessarily.
+  let firstTokenId: string | undefined
+
+  // eslint-disable-next-line no-unreachable-loop
+  for (const tokenId in allTokens) {
+    firstTokenId = tokenId
+    break
+  }
+  if (firstTokenId == null) return
+  const firstToken = allTokens[firstTokenId]
+
+  // Validate networkLocation input. Make sure the keys provided in networkLocation exist in
+  // the first token in the allTokens object.
+  for (const key in networkLocation) {
+    if (firstToken?.networkLocation?.[key] == null) {
+      console.warn(`findTokenIdByNetworkLocation: Unsupported networkLocation key: ${key}`)
+      return
+    }
+  }
+
+  for (const tokenId in allTokens) {
+    const edgeToken = allTokens[tokenId]
+    if (edgeToken?.networkLocation?.contractAddress?.toLowerCase() === networkLocation?.contractAddress?.toLowerCase()) {
+      return tokenId
+    }
```

Comment:
It looks like `contractAddress` is assumed to be the key within the query param `params.networkLocation`? Did you mean to do this?

If not, and instead you meant to use whatever keys are defined on `params.networkLocation`, then it seems like that could skip the whole "first key" validation song and dance:

```ts
export const findTokenIdByNetworkLocation = (params: FindTokenParams): EdgeTokenId | undefined => {
  const { networkLocation } = params
  let allTokens: EdgeTokenMap
  if ('allTokens' in params) {
    allTokens = params.allTokens
  } else {
    allTokens = params.account.currencyConfig[params.pluginId]?.allTokens
    if (allTokens == null) return
  }

  // eslint-disable-next-line no-labels
  outer: for (const tokenId in allTokens) {
    const edgeToken = allTokens[tokenId]
    for (const key in networkLocation) {
      const left = edgeToken?.networkLocation?.[key]
      if (typeof left !== 'string') {
        // Perhaps this should be an error throw?
        console.warn(`findTokenIdByNetworkLocation: Unsupported networkLocation key: ${key}`)
        return
      }
      const right = networkLocation[key]
      if (typeof right !== 'string') {
        // Perhaps this should be an error throw?
        console.warn(`findTokenIdByNetworkLocation: Unsupported networkLocation key: ${key}`)
        return
      }

      if (left.toLowerCase() !== right.toLowerCase()) {
        // eslint-disable-next-line no-labels
        continue outer
      }
    }
    // Found because it passed the checks in the inner loop
    return tokenId
  }
}
```

Maybe I misunderstand the function implementation?

Reply:
Good catch. Let me update as I had intended and just never completed.